### PR TITLE
build: force `npm install` after package.json changes

### PR DESCRIPTION
**Pull request by allisonkarlitskaya**
It's a reasonably common problem to pull, switch branches, or rebase on
top of new changes from master, and have changes to node module
dependencies which will cause the build to fail if `npm install` isn't
run again.  Even worse: sometimes the build succeeds, only to cause
runtime javascript errors during testing.

This is almost always indicated by a corresponding change in
package.json.

Make sure package-lock.json is newer than package.json, and if not,
display a warning about it, asking the user to fix it.  Use a ridiculous
amount of whitespace to make sure the message sticks out from the noise
and to make it easier to copy/paste the suggested command.

Also: add a new Makefile target `make npm` to explicitly touch
package-lock.json after a sucessful `npm install` (since npm won't do it
for itself if there are no changes).

### CODE REVIEW

File: Makefile.am [hunk 1]
```diff
@@ -224,6 +224,23 @@ clean-local::
 maintainer-clean-local::
 	rm -rf dist/ node_modules/ package-lock.json
 
+BUILT_SOURCES += package-lock.json
```
Comment (martinpitt) [inline Makefile.am]: Does that obsolete adding it to `EXTRA_DIST`? (It doesn't hurt, just curious)

File: Makefile.am [hunk 2]
```diff
@@ -224,6 +224,23 @@ clean-local::
...
+
+package-lock.json: package.json
+	@echo ''
+	@echo '    package-lock.json needs to be (re)generated'
+	@echo ''
+	@echo '    Please run:'
+	@echo ''
+	@echo '        make npm'
+	@echo ''
+	@false
```
Comment (martinpitt) [inline Makefile.am]: As discussed earlier, I think that strikes the right balance in preventing weird bugs and not accidentally hitting the network on `make check`.

File: Makefile.am [hunk 3]
```diff
@@ -224,6 +224,23 @@ clean-local::
...
+
+.PHONY: npm
+npm:
+	npm install
```
Comment (martinpitt) [inline Makefile.am]: I'm not totally happy about this, though. autogen.sh also does that, with a retry loop, and `npm prune`.  Plus the npm version check (although that may be obsolete now). How about one of these:

 * move the entire logic here, and change autogen.sh to call `make npm`
 * change the message to call autogen.sh
 * move the npm bits into tools/npm-install.sh and calling that from both here and autogen.sh

Comment (allisonkarlitskaya) [reply Makefile.am]: I like the `tools/` approach.  I'll do that.

File: Makefile.am (hunk 1 above)
Comment (allisonkarlitskaya) [reply Makefile.am]: No.  These are two different things (as discussed).

**Comment by martinpitt** [comment]
I like the approach! I'd just prefer to call the thing `node_modules/stamp`, that would DTRT with `rm -r node_modules/` and is less clutter in the root dir.

**Comment by allisonkarlitskaya** [comment]
> I like the approach! I'd just prefer to call the thing `node_modules/stamp`, that would DTRT with `rm -r node_modules/` and is less clutter in the root dir.

True on both counts, but I call it `.stamp` instead of `stamp`, just incase we randomly get some node module with that name in the future...

**Comment by martinpitt** [comment]
LOL -- I suppose we need to special-case that then :)

    SystemError: Could not determine license from cockpit-236.56.g0bed47f0a/node_modules/.stamp

**Review by martinpitt — APPROVED** [review]
This works nicely, many thanks for closing this ancient trap door!

**Pull request merged**